Add shell completion

emillon opened the pull request:
This provides a shell completion mechanism for dune. This relies on the bash completion API, which can be used with zsh as well.

The architecture is:

- `dune complete script` outputs a script to be sourced in the user's shell. It is comprised of a `_dune` function and the `complete -F _dune dune` command to register it. The `_dune` function can be used in cram tests to write natural-looking tests for this feature.
- this script calls `dune complete command` with the partial command-line. This internal command parses it to determine what the word being completed refers to: a command name, an argument name, or an argument value. The first two ones are part of the metadata `cmdliner` knows about; the last one is provided through a completion function that can be passed in one the `Arg` functions.
- the interface between `bash` and `dune complete command` is simple: it passes the command line and a position to complete at (this is necessary to encode the difference between `dune bui<tab>` and `dune build <tab>` for example), and reads an array from the output of the command.

The things I'm happy with:

- it is small!
- coverage is pretty good: command names, arguments (positional and optional, including optional arguments with optional names), and the `--` construct are supported. So, this is likely to improve the user experience already.
- it is easy to test through cram or unit tests (I chose the former).

Now, for the ugly bits...

- this effectively is a partial reimplementation of cmdliner inside `complete.ml`. If the exact parsing rules are different, it means that we can complete to something with different or wrong semantics.
- the vendored copy of cmdliner is patched to expose so that it is possible to use the private APIs. these two points need to be resolved before we can think about how to upstream this.
- some bits of the cmdliner API need to be modified to provide completion automatically. For example for things like `enum` it's easy to provide a completion function automatically.
- it is difficult to define the right API for the completion functions. `unit -> string list` is a first approximation but with some limitations. For example, getting a list of buildable targets needs to run under `Fiber`, but we can't pollute the API with it. Interestingly enough, algebraic effects seem like they would be an interesting solution for this.
- at the moment, we're not relying on the shell's completion helpers to complete things like filenames. To support this we would either need to implement that in OCaml, or extend the bash/dune interface so that the completion function could call `compgen -f` based on the dune output.
- as a way to tie the two previous points: if we wanted to complete `dune build dir/file<tab>`, it would be a lot more efficient to pass the prefix to the build system and let it compute just the targets that match this, rather than compute everything and filter it afterwards. So that prefix would need to appear in the completion API.

emillon committed:
Add shell completion

This provides a shell completion mechanism for dune. This relies on the
bash completion API, which can be used with zsh as well.

The architecture is:

- `dune complete script` outputs a script to be sourced in the user's
  shell. It is comprised of a `_dune` function and the `complete -F
  _dune dune` command to register it. The `_dune` function can be used
  in cram tests to write natural-looking tests for this feature.
- this script calls `dune complete command` with the partial
  command-line. This internal command parses it to determine what the
  word being completed refers to: a command name, an argument name, or
  an argument value. The first two ones are part of the metadata
  `cmdliner` knows about; the last one is provided through a completion
  function that can be passed in one the `Arg` functions.
- the interface between `bash` and `dune complete command` is simple:
  it passes the command line and a position to complete at (this is
  necessary to encode the difference between `dune bui<tab>` and `dune
  build <tab>` for example), and reads an array from the output of the
  command.

The things I'm happy with:

- it is small!
- coverage is pretty good: command names, arguments (positional and
  optional, including optional arguments with optional names), and the
  `--` construct are supported. So, this is likely to improve the user
  experience already.
- it is easy to test through cram or unit tests (I chose the former).

Now, for the ugly bits...

- this effectively is a partial reimplementation of cmdliner inside
  `complete.ml`. If the exact parsing rules are different, it means that
  we can complete to something with different or wrong semantics.
- the vendored copy of cmdliner is patched to expose so that it is
  possible to use the private APIs. these two points need to be resolved
  before we can think about how to upstream this.
- some bits of the cmdliner API need to be modified to provide
  completion automatically. For example for things like `enum` it's easy
  to provide a completion function automatically.
- it is difficult to define the right API for the completion functions.
  `unit -> string list` is a first approximation but with some
  limitations. For example, getting a list of buildable targets needs to
  run under `Fiber`, but we can't pollute the API with it. Interestingly
  enough, algebraic effects seem like they would be an interesting
  solution for this.
- at the moment, we're not relying on the shell's completion helpers to
  complete things like filenames. To support this we would either need
  to implement that in OCaml, or extend the bash/dune interface so that
  the completion function could call `compgen -f` based on the dune
  output.
- as a way to tie the two previous points: if we wanted to complete
  `dune build dir/file<tab>`, it would be a lot more efficient to pass
  the prefix to the build system and let it compute just the targets
  that match this, rather than compute everything and filter it
  afterwards. So that prefix would need to appear in the completion API.

Signed-off-by: Etienne Millon <me@emillon.org>

emillon commented:
Some random thoughts on API design while I'm rebasing the next iteration:
- converters seems like the right place to add a completion function. there's probably be a way to unify the parser and completion parts but it's a bit quirky. The good candidates for a completion function are either finite things (`enum`, `bool`) or nested/structured things (paths) or combinations thereof (BIP39 / what3words). Things like parsing with derivatives are tempting but they just tell us that when parsing an `int`, the only characters accepted are digits (well, a leading `0` could suggest `x` or `b` maybe). We can't use that info on the command line.
- I was looking for ways to hook into the shell's helpers like `compgen -F` (complete filenames) and for example return a variant that gets turned either into a list of words or the right `compgen` call. That's actually not great for composition. Consider the completion command for `pair enum file`: to complete `a,` (where `a` is an `enum`), we'd need to compute '["a," ^ b for b in compgen -F]` on the shell side. obviously not great. for that reason, I think that `string list` is the best we can do, and that tools like `file` have to be implemented in ocaml.
- sometimes, the ability to complete depends on the partial input. Consider again `pair a b` where `a` does not have a completion function, but `b` does. We can complete `something,` (using b's completion function) but we can't complete the empty string. So "optional completion function" might be more subtle than just `(unit -> string list) option` (if it were the case, `pair a b` would be a `Some _`, but it'd have to return `[]` in the case of the empty string.

rgrinberg commented:
How often do we use pair in dune anyway? I can't think of any cases off the top of my head.

Should we perhaps not handle it properly for now?

emillon commented:
Yes, you're right that we probably don't use pair (we use list though for `-p` but it looks like it's easier to support - I expected the opposite). I'm trying to move most of the code to cmdliner so the API needs to be at least compatible, but you're right it does not have to be a priority as long as we can get something that works.

emillon commented:
Alright, this is getting better:
- I moved a bunch of completion code to cmdliner
- completion hook now gets the prefix as argument
- implemented list completion
- this is now completing simple cases of target names. I'll have to implement the suggestion described at https://github.com/ocaml/dune/blob/61c0585f8c85195979d4d2ecbb6f94ce8c73a1b4/bin/target.ml#L28-L30 in order to be usable in an actual project.

emillon committed:
Complete aliases

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Remove dune complete target

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Add test with = syntax

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Complete @install

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Load alias names

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Complete dirs

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Handle --arg=value syntax

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
pass --display quiet when completing

emillon committed:
Completion of targets

Signed-off-by: Etienne Millon <me@emillon.org>

Alizter commented on bin/common.ml:
```diff
@@ -1142,9 +1142,9 @@ let context_arg ~doc =
   let info = Cmd.info "term-default" in
   let cmd = Cmd.v info term in
-  match Cmd.eval_value ~argv:[| Sys.argv.(0) |] cmd with
+  match Cmd.eval_value ~argv:[| Sys.argv.(0); "--display=quiet" |] cmd with
```
You can do this internally in Build_cmd.complete now.

emillon replied:
What do you mean? I can change `display` but it's still not possible to get a `Common.t` value without parsing a CLI.

emillon committed:
Add an intermediate state for more stability

This adds an extra indirection layer:

- the end user will call `dune complete setup`. the output of this script is supposed to be pretty stable and can be copied in dotfiles.
- it generates a call to `dune complete script` which outputs the shell/dune FFI. The output of this command is not stable.
- in the current implementation, this contains a call to `dune complete
  command` which passes the full CLI and a `--position` argument

Signed-off-by: Etienne Millon <me@emillon.org>

Alizter commented on bin/target.ml:
```diff
@@ -104,15 +104,19 @@ let target_hint (_setup : Dune_rules.Main.build_system) path =
         else None)
   in
-  let candidates = String.Set.of_list candidates |> String.Set.to_list in
+  String.Set.of_list candidates |> String.Set.to_list
```
Is this to remove dupes?

emillon replied:
yes

rgrinberg commented:
Please re-open once you plan to re-visit this.

Alizter commented:
This has been subsumed by #14779. Support for various subcommands is on-going and tracked in #126, #15335, #15336, #7254  etc.
